src/metabase/lib/metadata.cljc

```diff
@@ -31,23 +44,110 @@
    [:id ::lib.schema.id/table]
    [:name ::lib.schema.common/non-blank-string]
    [:schema {:optional true} [:maybe ::lib.schema.common/non-blank-string]]
-   [:fields [:sequential ColumnMetadata]]])
+   ;; This is now optional! If the [[DatabaseMetadataProvider]] provides it, great, but if not we can always make the
+   ;; subsequent request to fetch fields separately.
+   [:fields {:optional true} [:sequential ColumnMetadata]]])
 
 (def DatabaseMetadata
   "Malli schema for the DatabaseMetadata as returned by `GET /api/database/:id/metadata` -- what should be available to
   the frontend Query Builder."
   [:map
    [:lib/type [:= :metadata/database]]
    [:id ::lib.schema.id/database]
-   [:tables [:sequential TableMetadata]]])
+   ;; Like `:fields` for [[TableMetadata]], this is now optional -- we can fetch the Tables separately if needed.
+   [:tables {:optional true} [:sequential TableMetadata]]])
 
-{:database 1
- :type     :pipeline
- :stages   [{:lib/type :mbql.stage/mbql
-             :native   "SELECT id, name FROM VENUES;"}]}
+(def DatabaseMetadataProvider
+  "Schema for something that satisfies the [[lib.metadata.protocols/DatabaseMetadataProvider]] protocol."
+  [:fn lib.metadata.protocols/DatabaseMetadataProvider?])
 
-{:columns [{:name "id", :base-type :type/Integer}
-           {:name "name", :base-type :type/Text}]}
+(defmulti ^:private ->database-metadata-provider*
+  {:arglists '([x])}
+  lib.dispatch/dispatch-value)
+
+(defmethod ->database-metadata-provider* :default
+  [x]
+  x)
```

Comment:
That's why I did it, actually... `->database-metadata-provider` is an `mu/defn` so it enforces the `DatabaseMetadataProvider` schema for its output. The idea is that you should use `->database-metadata-provider`, but implement `->database-metadata-provider*` if we wanted to add support for more stuff being converted to metadata providers in the future. Maybe there's some better way to do that tho. It might be too easy to accidentally use `->database-metadata-provider*` and lose the checking